### PR TITLE
ossl_namemap_name2_num: Avoid unnecessary OPENSSL_strndup().

### PR DESCRIPTION
**Pull request by t8m**
Related to https://github.com/openssl/openssl/issues/16791#issuecomment-1129884672

### CODE REVIEW

File: crypto/core_namemap.c [hunk 1]
```diff
@@ -195,18 +192,24 @@ int ossl_namemap_name2num_n(const OSSL_NAMEMAP *namemap,
...
 
-int ossl_namemap_name2num(const OSSL_NAMEMAP *namemap, const char *name)
+int ossl_namemap_name2num_n(const OSSL_NAMEMAP *namemap,
+                            const char *name, size_t name_len)
 {
-    if (name == NULL)
+    char *tmp;
+    int ret;
+
+    if (name == NULL || (tmp = OPENSSL_strndup(name, name_len)) == NULL)
```
Comment (paulidale) [inline crypto/core_namemap.c:207]: Perfect use case for C99's dynamic arrays :)

File: crypto/core_namemap.c [hunk 2]
```diff
@@ -241,14 +244,20 @@ static int namemap_add_name_n(OSSL_NAMEMAP *namemap, int number,
 {
     NAMENUM_ENTRY *namenum = NULL;
     int tmp_number;
+    char *tmp = OPENSSL_strndup(name, name_len);
```
Comment (paulidale) [inline crypto/core_namemap.c:247]: Maybe add a similar comment to this function too:
```c
/* This function is not thread safe, the namemap must be locked */
```

**Review by paulidale — APPROVED** [review]
Looks good!  Tracking the locking was a pain, hence the suggestions to add the comments.

File: crypto/core_namemap.c [hunk 3]
```diff
@@ -312,13 +322,22 @@ int ossl_namemap_add_names(OSSL_NAMEMAP *namemap, int number,
      */
     for (p = names; *p != '\0'; p = (q == NULL ? p + l : q + 1)) {
         int this_number;
+        char *allocated;
```
Comment (slontis) [inline crypto/core_namemap.c:325]: Would doing a single strdup of names before the loop any better here?

Comment (paulidale) [reply crypto/core_namemap.c:325]: Probably.

Comment (t8m) [reply crypto/core_namemap.c:325]: Good point, I'll do this.

**Comment by t8m** [comment]
@paulidale @beldmit I've added further simplification by dropping the ossl_namemap_add_name_n() which is not used anywhere in the codebase. That allowed simplifying ossl_namemap_add_names().

Please re-approve.

File: crypto/core_namemap.c [hunk 4]
```diff
@@ -274,92 +281,92 @@ int ossl_namemap_add_name_n(OSSL_NAMEMAP *namemap, int number,
...
         int this_number;
 
-        if ((q = strchr(p, separator)) == NULL)
-            l = strlen(p);       /* offset to \0 */
-        else
-            l = q - p;           /* offset to the next separator */
+        q = p + strlen(p) + 1;
 
-        this_number = namemap_add_name_n(namemap, number, p, l);
+        this_number = namemap_add_name(namemap, number, p);
```
Comment (FdaSilvaYY) [inline crypto/core_namemap.c:354]: Same here, you already know the name len.

File: crypto/core_namemap.c [hunk 5]
```diff
@@ -236,18 +240,21 @@ const char *ossl_namemap_num2name(const OSSL_NAMEMAP *namemap, int number,
...
+/* This function is not thread safe, the namemap must be locked */
+static int namemap_add_name(OSSL_NAMEMAP *namemap, int number,
+                            const char *name)
 {
     NAMENUM_ENTRY *namenum = NULL;
     int tmp_number;
 
     /* If it already exists, we don't add it */
-    if ((tmp_number = namemap_name2num_n(namemap, name, name_len)) != 0)
+    if ((tmp_number = namemap_name2num(namemap, name)) != 0)
```
[the remark this answers is not in the document]
Comment (FdaSilvaYY) [reply crypto/core_namemap.c:251]: I fixed my typo/paste error in this comment

**Comment by t8m** [comment]
> Some suggestions. Feel free to consider them

I prefer keeping the code as is.

**Review by paulidale — APPROVED** [review]

File: crypto/core_namemap.c [hunk 6]
```diff
@@ -274,92 +281,92 @@ int ossl_namemap_add_name_n(OSSL_NAMEMAP *namemap, int number,
...
+            q = p + l;
+        } else {
             l = q - p;           /* offset to the next separator */
+            *q++ = '\0';
+        }
 
-        this_number = namemap_name2num_n(namemap, p, l);
-
-        if (*p == '\0' || *p == separator) {
+        if (*p == '\0') {
```
Comment (slontis) [inline crypto/core_namemap.c:328]: How would it get inside this block?

Comment (paulidale) [reply crypto/core_namemap.c:328]: See line 325 :)

Comment (slontis) [reply crypto/core_namemap.c:328]: ok..

**Comment by t8m** [comment]
ping for the second review

**Comment by beldmit** [comment]
Still approved

**Comment by beldmit** [comment]
Merged. Thanks!

**Pull request closed without merging**